state/protocol/badger/mutator.go

```diff
@@ -658,45 +620,204 @@ func (m *FollowerState) Finalize(ctx context.Context, blockID flow.Identifier) e
 }
 
-// epochStatus computes the EpochStatus for the given block
-// BEFORE applying the block payload itself
+// epochFallbackTriggeredByFinalizedBlock checks whether finalizing the input block
+// would trigger epoch emergency fallback mode. In particular, we trigger epoch
+// fallback mode when finalizing a block B when:
+// 1. B is the first finalized block with view greater than or equal to the epoch
+//    commitment deadline for the current epoch.
+// 2. The next epoch has not been committed as of B.
+//
+// This function should only be called when epoch fallback *has not already been triggered*.
+// See protocol.Params for more details on the epoch commitment deadline.
+//
+// No errors are expected during normal operation.
+func (m *FollowerState) epochFallbackTriggeredByFinalizedBlock(block *flow.Header, epochStatus *flow.EpochStatus, currentEpochSetup *flow.EpochSetup) (bool, error) {
+
+	// 1. determine whether block B is past the epoch commitment deadline
+	safetyThreshold, err := m.Params().EpochCommitSafetyThreshold()
+	if err != nil {
+		return false, fmt.Errorf("could not get epoch commit safety threshold: %w", err)
+	}
+	epochCommitmentDeadline := currentEpochSetup.FinalView - safetyThreshold
+	blockExceedsDeadline := block.View >= epochCommitmentDeadline
+
+	// 2. determine whether the next epoch is committed w.r.t. block B
+	currentEpochPhase, err := epochStatus.Phase()
+	if err != nil {
+		return false, fmt.Errorf("could not get current epoch phase: %w", err)
+	}
+	isNextEpochCommitted := currentEpochPhase == flow.EpochPhaseCommitted
+
+	blockTriggersEpochFallback := blockExceedsDeadline && !isNextEpochCommitted
+	return blockTriggersEpochFallback, nil
+}
+
+// epochTransitionMetricsAndEventsOnBlockFinalized determines metrics to update
+// and protocol events to emit, if this block is the first of a new epoch.
+//
+// Protocol events and updating metrics happen when we finalize the _first_
+// block of the new Epoch (same convention as for Epoch-Phase-Changes).
+// Approach: We retrieve the parent block's epoch information.
+// When this block's view exceeds the parent epoch's final view, this block
+// represents the first block of the next epoch. Therefore we update metrics
+// related to the epoch transition here.
+//
+// This function should only be called when epoch fallback *has not already been triggered*.
+// No errors are expected during normal operation.
+func (m *FollowerState) epochTransitionMetricsAndEventsOnBlockFinalized(block *flow.Header, currentEpochSetup *flow.EpochSetup) (
+	metrics []func(),
+	events []func(),
+	err error,
+) {
+
+	parentBlocksEpoch := m.AtBlockID(block.ParentID).Epochs().Current()
+	parentEpochFinalView, err := parentBlocksEpoch.FinalView()
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not get parent epoch final view: %w", err)
+	}
+
+	if block.View > parentEpochFinalView {
+		events = append(events, func() { m.consumer.EpochTransition(currentEpochSetup.Counter, block) })
```

Comment:
Would it make sense to add a sanity check here to verify that `currentEpochSetup` is actually for epoch `parentBlocksEpoch.Counter()+1`? This check has marginal runtime overhead and would catch already a broad class of bugs.

Reply:
👍 Good idea, added in https://github.com/onflow/flow-go/pull/2625/commits/b74d85fbc7b799d808b86770e275ae9b25e64657